File: code/ch03/3.4.1.longestPalindrome.java
import java.util.*;
import java.lang.Math;

public class Solution {
  public String extend(int i, int j, String s) {
    while (i >= 0 && j < s.length() && s.charAt(i) ==
        s.charAt(j)) {
      i -= 1;
      j += 1;
    }
    return s.substring(i + 1, j);
  }
  public String longestPalindrome(String s) {
    int n = s.length();
    if (n == 0) return "";
    String res = s.substring(0, 1);
    for (int i = 0; i < n - 1; i++) {
      String e1 = extend(i, i, s);
      String e2 = extend(i, i + 1, s);
      if (Math.max(e1.length(), e2.length()) > res.length())
        res = ((e1.length() > e2.length()) ? e1 : e2);
    }
    return res;
  }
}

File: code/ch03/3.4.1.longestPalindrome.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n == 0:
            return ""
        res = s[0]

        def extend(i: int, j: int, s: str) -> str:
            while i >= 0 and j < len(s) and s[i] == s[j]:
                i -= 1
                j += 1
            return s[i + 1 : j]

        for i in range(n - 1):
            # 以自身为中心点
            e1 = extend(i, i, s)
            # 以自身和自身的下一个元素为中心点
            e2 = extend(i, i + 1, s)
            if max(len(e1), len(e2)) > len(res):
                res = e1 if len(e1) > len(e2) else e2
        return res
